spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/pubsub/health/PubSubHealthIndicator.java

```diff
@@ -16,55 +16,66 @@
 
 package com.google.cloud.spring.autoconfigure.pubsub.health;
 
-import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
 
-import com.google.api.gax.rpc.ApiException;
 import com.google.api.gax.rpc.StatusCode;
-import com.google.api.gax.rpc.StatusCode.Code;
-import com.google.cloud.spring.pubsub.core.PubSubTemplate;
 
 import org.springframework.boot.actuate.health.AbstractHealthIndicator;
 import org.springframework.boot.actuate.health.Health;
 import org.springframework.util.Assert;
 
 /**
- * Default implemenation of
+ * Default implementation of
  * {@link org.springframework.boot.actuate.health.HealthIndicator} for Pub/Sub. Validates
- * if connection is successful by looking for a random generated subscription name that
- * won't be found. If no subscription is found we know the client is able to connect to
- * GCP Pub/Sub APIs.
+ * if connection is successful by pulling message asynchronously from the pubSubHealthTemplate.
+ *
+ * If a custom subscription has been specified, this health indicator will only signal up
+ * if messages are successfully pulled and acknowledged.
+ *
+ * If no subscription has been specified, this health indicator will pull messages from a random subscription
+ * that is expected not to exist. It will signal up if it is able to connect to GCP Pub/Sub APIs,
+ * i.e. the pull results in a response of {@link StatusCode.Code#NOT_FOUND} or
+ * {@link StatusCode.Code#PERMISSION_DENIED}.
  *
  * @author Vinicius Carvalho
+ * @author Patrik Hörlin
  *
  * @since 1.2.2
  */
 public class PubSubHealthIndicator extends AbstractHealthIndicator {
 
-	private final PubSubTemplate pubSubTemplate;
+	private final PubSubHealthTemplate pubSubHealthTemplate;
 
-	public PubSubHealthIndicator(PubSubTemplate pubSubTemplate) {
+	public PubSubHealthIndicator(PubSubHealthTemplate pubSubHealthTemplate) {
```

Comment:
Since `PubSubHealthTemplate` is now a package-private class, we can't make it part of the public interface here. So, we'll have to accept a bean of `PubSubHealthTemplate` like before and just create the `PubSubHealthTemplate` internally. Alternatively, you can just move the methods from the `PubSubHealthTemplate` class into this class and remove `PubSubHealthTemplate` altogether.